clang/lib/Headers/hlsl/hlsl_intrinsic_helpers.h

```diff
@@ -71,6 +71,23 @@ constexpr vector<T, L> reflect_vec_impl(vector<T, L> I, vector<T, L> N) {
 #endif
 }
 
+template <typename T> constexpr T refract_impl(T I, T N, T Eta) {
+  T K = 1 - Eta * Eta * (1 - (N * I * N * I));
+  T Result = (Eta * I - (Eta * N * I + sqrt(K)) * N);
+  return select<T>(K < 0, static_cast<T>(0), Result);
+}
+
+template <typename T, int L>
+constexpr vector<T, L> refract_vec_impl(vector<T, L> I, vector<T, L> N, T Eta) {
+#if (__has_builtin(__builtin_spirv_refract))
+  return __builtin_spirv_refract(I, N, Eta);
+#else
+  vector<T, L> K = 1 - Eta * Eta * (1 - dot(N, I) * dot(N, I));
+  vector<T, L> Result = (Eta * I - (Eta * dot(N, I) + sqrt(K)) * N);
+  return select<vector<T, L>>(K < 0, vector<T, L>(0), Result);
```

Comment:
Our most recent pattern on these has been to do one function when possible instead of a scalar and vector implementation. I beleive this is a case where that might be possible. Can you experiment with the following?

In `hlsl_detail.h` add
```cpp
template <typename T>
struct is_vector {
  static const bool value = false;
};

/*NOTE: (don't include this comment) what I am doing here is adding a specialization for vector<T, N>*/
template <typename T, int N>
struct is_vector<vector<T, N>> {
  static const bool value = true;
};
```

Then line 82 changes to 
```suggestion
#if (__has_builtin(__builtin_spirv_refract) && is_vector<T>)
  return __builtin_spirv_refract(I, N, Eta);
#else
  T K = 1 - Eta * Eta * (1 - dot(N, I) * dot(N, I));
  T Result = (Eta * I - (Eta * dot(N, I) + sqrt(K)) * N);
  return select(K < 0, 0, Result);
```